File: GBandRN/android/app/src/main/java/com/gbandrn/GBandModule.kt
package com.gbandrn

import android.os.Handler
import android.os.Looper
import com.facebook.react.bridge.*
import com.facebook.react.modules.core.DeviceEventManagerModule
import com.inuker.bluetooth.library.search.SearchResult
import com.inuker.bluetooth.library.search.response.SearchResponse
import com.inuker.bluetooth.library.Code
import com.inuker.bluetooth.library.model.BleGattProfile
import com.inuker.bluetooth.library.utils.BluetoothUtils
import com.veepoo.protocol.VPOperateManager
import com.veepoo.protocol.listener.base.IConnectResponse
import com.veepoo.protocol.listener.base.INotifyResponse
import com.veepoo.protocol.listener.data.*
import com.veepoo.protocol.model.datas.*

class GBandModule(private val reactContext: ReactApplicationContext) :
    ReactContextBaseJavaModule(reactContext) {

    override fun getName() = "GBandModule"

    private val mainHandler = Handler(Looper.getMainLooper())

    // ── Measurement state (SDK is serial — one at a time) ─────────────────
    private var isMeasuringHeart    = false
    private var isMeasuringSpo2     = false
    private var isMeasuringBp       = false
    private var isMeasuringTemp     = false
    private var isWellnessRunning   = false

    // ── Event emitter helper ──────────────────────────────────────────────
    private fun emit(event: String, data: Any?) {
        reactContext
            .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
            .emit(event, data)
    }

    private fun emitMap(event: String, block: WritableMap.() -> Unit) {
        val map = Arguments.createMap()
        map.block()
        emit(event, map)
    }

    // ── Scan ──────────────────────────────────────────────────────────────
    @ReactMethod
    fun startScan(promise: Promise) {
        try {
            if (!BluetoothUtils.isBluetoothEnabled()) {
                promise.reject("BT_DISABLED", "Bluetooth is not enabled")
                return
            }
            VPOperateManager.getInstance().stopScanDevice()
            VPOperateManager.getInstance().disconnectWatch {}
            VPOperateManager.getInstance().startScanDevice(object : SearchResponse {
                override fun onSearchStarted() {
                    emitMap("onScanStatus") { putString("status", "scanning") }
                }
                override fun onDeviceFounded(device: SearchResult) {
                    emitMap("onDeviceFound") {
                        putString("name", device.name ?: "Unknown")
                        putString("mac", device.address)
                        putInt("rssi", device.rssi)
                    }
                }
                override fun onSearchStopped() {
                    emitMap("onScanStatus") { putString("status", "stopped") }
                }
                override fun onSearchCanceled() {
                    emitMap("onScanStatus") { putString("status", "canceled") }
                }
            })
            // Auto-stop after 12s
            mainHandler.postDelayed({ VPOperateManager.getInstance().stopScanDevice() }, 12000)
            promise.resolve(null)
        } catch (e: Exception) {
            promise.reject("SCAN_ERR", e.message)
        }
    }

    @ReactMethod
    fun stopScan(promise: Promise) {
        VPOperateManager.getInstance().stopScanDevice()
        promise.resolve(null)
    }

    // ── Connect ───────────────────────────────────────────────────────────
    @ReactMethod
    fun connectDevice(mac: String, name: String, promise: Promise) {
        try {
            VPOperateManager.getInstance().stopScanDevice()
            VPOperateManager.getInstance().connectDevice(
                mac, name,
                IConnectResponse { code, _, _ ->
                    if (code == Code.REQUEST_SUCCESS) {
                        emitMap("onConnectionState") { putString("state", "ble_connected") }
                    } else {
                        emitMap("onConnectionState") {
                            putString("state", "failed")
                            putInt("code", code)
                        }
                    }
                },
                INotifyResponse { state ->
                    if (state == Code.REQUEST_SUCCESS) {
                        mainHandler.postDelayed({ verifyPassword(mac, name) }, 1000)
                    }
                }
            )
            promise.resolve(null)
        } catch (e: Exception) {
            promise.reject("CONNECT_ERR", e.message)
        }
    }

    private fun verifyPassword(mac: String, name: String) {
        VPOperateManager.getInstance().confirmDevicePwd(
            {},
            IPwdDataListener { pwdData ->
                val status = pwdData.mStatus?.toString() ?: ""
                if (status.contains("SUCESS") || status.contains("SUCCESS")) {
                    emitMap("onConnected") {
                        putString("mac", mac)
                        putString("name", name)
                        putString("firmware", pwdData.deviceVersion ?: "")
                    }
                } else {
                    emitMap("onConnectionState") {
                        putString("state", "auth_failed")
                        putString("reason", status)
                    }
                }
            },
            object : IDeviceFuctionDataListener {
                override fun onFunctionSupportDataChange(caps: FunctionDeviceSupportData) {
                    emitMap("onCapabilities") {
                        putBoolean("bp",           caps.isSupportBp)
                        putBoolean("spo2",         caps.isSupportSpoH)
                        putBoolean("temp",         caps.isSupportTemperatureFunction)
                        putBoolean("ecg",          caps.isSupportECG)
                        putBoolean("hrv",          caps.isSupportHRV)
                        putBoolean("precisionSleep", caps.isSupportPrecisionSleep)
                        putBoolean("stress",       caps.isSupportStressDetect)
                        putBoolean("respRate",     caps.isSupportBeathFunction)
                        putBoolean("bloodGlucose", caps.isSupportBloodGlucose)
                    }
                }
                override fun onDeviceFunctionPackage1Report(d: DeviceFunctionPackage1?) {}
                override fun onDeviceFunctionPackage2Report(d: DeviceFunctionPackage2?) {}
                override fun onDeviceFunctionPackage3Report(d: DeviceFunctionPackage3?) {}
                override fun onDeviceFunctionPackage4Report(d: DeviceFunctionPackage4?) {}
                override fun onDeviceFunctionPackage5Report(d: DeviceFunctionPackage5?) {}
            },
            null, null, "0000", false
        )
    }

    @ReactMethod
    fun disconnect(promise: Promise) {
        VPOperateManager.getInstance().disconnectWatch {
            emitMap("onConnectionState") { putString("state", "disconnected") }
        }
        promise.resolve(null)
    }

    // ── Battery ───────────────────────────────────────────────────────────
    @ReactMethod
    fun readBattery(promise: Promise) {
        VPOperateManager.getInstance().readBattery({}) { data ->
            val pct = if (data.isPercent) data.batteryPercent else data.batteryLevel * 25
            promise.resolve(pct)
        }
    }

    // ── Steps / Distance / Calories ───────────────────────────────────────
    @ReactMethod
    fun readSteps(promise: Promise) {
        VPOperateManager.getInstance().readSportStep({}) { data ->
            val map = Arguments.createMap()
            map.putInt("steps", data.step)
            map.putDouble("distanceKm", data.dis.toDouble())
            map.putDouble("caloriesKcal", data.kcal.toDouble())
            promise.resolve(map)
        }
    }

    // ── Sleep ─────────────────────────────────────────────────────────────
    @ReactMethod
    fun readSleep(promise: Promise) {
        VPOperateManager.getInstance().readSleepData(
            {},
            object : ISleepDataListener {
                override fun onSleepDataChange(day: String, data: SleepData) {
                    val map = Arguments.createMap()
                    map.putInt("totalMinutes",  data.allSleepTime)
                    map.putInt("deepMinutes",   data.deepSleepTime)
                    map.putInt("lightMinutes",  data.lowSleepTime)
                    map.putInt("wakeCount",     data.wakeCount)
                    map.putInt("sleepQuality",  data.sleepQulity)
                    promise.resolve(map)
                }
                override fun onSleepProgress(p: Float) {}
                override fun onSleepProgressDetail(day: String, pkg: Int) {}
                override fun onReadSleepComplete() {}
            },
            1
        )
    }

    // ── OriginData3 (HR history, BP history, SpO2 history, Resp, HRV) ─────
    @ReactMethod
    fun readOriginData(promise: Promise) {
        VPOperateManager.getInstance().readOriginData(
            { _ -> },
            object : IOriginData3Listener {
                private var lastHr = -1
                private var lastSys = -1
                private var lastDia = -1
                private var lastResp = -1
                private var lastOxy = -1

                override fun onOriginData3Change(date: String, dataList: List<OriginData3>?) {
                    dataList?.forEach { d3 ->
                        if (d3.rateValue > 0) lastHr = d3.rateValue
                        if (d3.highValue > 0) lastSys = d3.highValue
                        if (d3.lowValue > 0)  lastDia = d3.lowValue
                        d3.resRates?.forEach { r -> if (r > 0) lastResp = r }
                        d3.oxygens?.forEach { o -> if (o > 0) lastOxy = o }
                    }
                }

                override fun onOriginHRVOriginListDataChange(list: List<HRVOriginData>?) {
                    list?.lastOrNull()?.let { last ->
                        val hrv = last.hrv
                        if (hrv in 1..254) {
                            emitMap("onHRV") { putInt("hrv", hrv) }
                        }
                    }
                }

                override fun onOringinHalfHourDataChange(d: OriginHalfHourData?) {}
                override fun onReadOriginProgressDetail(day: Int, date: String, all: Int, curr: Int) {}
                override fun onReadOriginProgress(p: Float) {}

                override fun onReadOriginComplete() {
                    val map = Arguments.createMap()
                    if (lastHr   > 0) map.putInt("heartRate", lastHr)
                    if (lastSys  > 0) map.putInt("bpSystolic", lastSys)
                    if (lastDia  > 0) map.putInt("bpDiastolic", lastDia)
                    if (lastResp > 0) map.putInt("respirationRate", lastResp)
                    if (lastOxy  > 0) map.putInt("spo2", lastOxy)
                    promise.resolve(map)
                }
            },
            1
        )
    }

    // ── Heart Rate live measurement ───────────────────────────────────────
    @ReactMethod
    fun startHeartMeasure(promise: Promise) {
        if (isMeasuringHeart) { promise.reject("BUSY", "Already measuring"); return }
        isMeasuringHeart = true
        VPOperateManager.getInstance().startDetectHeart({}) { data ->
            val bpm = data.rateValue
            if (bpm > 0) emitMap("onHeartRate") { putInt("bpm", bpm) }
        }
        promise.resolve(null)
    }

    @ReactMethod
    fun stopHeartMeasure(promise: Promise) {
        isMeasuringHeart = false
        VPOperateManager.getInstance().stopDetectHeart {}
        promise.resolve(null)
    }

    // ── SpO2 live measurement ─────────────────────────────────────────────
    @ReactMethod
    fun startSpo2Measure(promise: Promise) {
        if (isMeasuringSpo2) { promise.reject("BUSY", "Already measuring"); return }
        isMeasuringSpo2 = true
        VPOperateManager.getInstance().startDetectSPO2H({}) { data ->
            val o2 = data.oxygen
            if (o2 > 0) emitMap("onSpo2") { putInt("value", o2) }
        }
        promise.resolve(null)
    }

    @ReactMethod
    fun stopSpo2Measure(promise: Promise) {
        isMeasuringSpo2 = false
        VPOperateManager.getInstance().stopDetectSPO2H {}
        promise.resolve(null)
    }

    // ── Blood Pressure live measurement ───────────────────────────────────
    @ReactMethod
    fun startBpMeasure(promise: Promise) {
        if (isMeasuringBp) { promise.reject("BUSY", "Already measuring"); return }
        isMeasuringBp = true
        VPOperateManager.getInstance().startDetectBP({}) { data ->
            val hi = data.highValue; val lo = data.lowValue
            if (hi > 0 && lo > 0) emitMap("onBP") {
                putInt("systolic", hi)
                putInt("diastolic", lo)
            }
        }
        promise.resolve(null)
    }

    @ReactMethod
    fun stopBpMeasure(promise: Promise) {
        isMeasuringBp = false
        VPOperateManager.getInstance().stopDetectBP {}
        promise.resolve(null)
    }

    // ── Body Temperature ──────────────────────────────────────────────────
    @ReactMethod
    fun startTempMeasure(promise: Promise) {
        if (isMeasuringTemp) { promise.reject("BUSY", "Already measuring"); return }
        isMeasuringTemp = true
        VPOperateManager.getInstance().startDetectTempture({}) { data ->
            val temp = data.tempture
            if (temp > 0f) emitMap("onTemperature") { putDouble("celsius", temp.toDouble()) }
        }
        promise.resolve(null)
    }

    @ReactMethod
    fun stopTempMeasure(promise: Promise) {
        isMeasuringTemp = false
        VPOperateManager.getInstance().stopDetectTempture {}
        promise.resolve(null)
    }

    // ── Wellness Check (Stress + Fatigue + HRV via MiniCheckup) ──────────
    @ReactMethod
    fun startWellnessCheck(promise: Promise) {
        if (isWellnessRunning) { promise.reject("BUSY", "Already measuring"); return }
        isWellnessRunning = true
        VPOperateManager.getInstance().startMiniCheckup({}) { result ->
            val map = Arguments.createMap()
            if (result.stressLevel >= 0)   map.putInt("stress",      result.stressLevel)
            if (result.fatigueDegree >= 0) map.putInt("fatigue",     result.fatigueDegree)
            if (result.hrv in 1..254)      map.putInt("hrv",         result.hrv)
            if (result.heartRate > 0)      map.putInt("heartRate",   result.heartRate)
            if (result.oxygen > 0)         map.putInt("spo2",        result.oxygen)
            if (result.temperature > 0f)   map.putDouble("temp",     result.temperature.toDouble())
            if (result.systolic > 0)       map.putInt("bpSystolic",  result.systolic)
            if (result.diastolic > 0)      map.putInt("bpDiastolic", result.diastolic)
            emitMap("onWellnessResult") {
                if (result.stressLevel >= 0)   putInt("stress",      result.stressLevel)
                if (result.fatigueDegree >= 0) putInt("fatigue",     result.fatigueDegree)
                if (result.hrv in 1..254)      putInt("hrv",         result.hrv)
                if (result.heartRate > 0)      putInt("heartRate",   result.heartRate)
                if (result.oxygen > 0)         putInt("spo2",        result.oxygen)
                if (result.temperature > 0f)   putDouble("temp",     result.temperature.toDouble())
                if (result.systolic > 0)       putInt("bpSystolic",  result.systolic)
                if (result.diastolic > 0)      putInt("bpDiastolic", result.diastolic)
            }
        }
        promise.resolve(null)
    }

    @ReactMethod
    fun stopWellnessCheck(promise: Promise) {
        isWellnessRunning = false
        VPOperateManager.getInstance().stopMiniCheckup {}
        promise.resolve(null)
    }

    // ── Required override for RN event emitter ────────────────────────────
    @ReactMethod
    fun addListener(eventName: String) {}

    @ReactMethod
    fun removeListeners(count: Int) {}
}

File: GBandRN/ios/GBandRN/GBandModule.swift
import Foundation
import CoreBluetooth

// VeepooBleSDK imports — framework must be in ios/Frameworks/VeepooBleSDK.framework
// Download from: https://github.com/HBandSDK/iOS_Ble_SDK
@_implementationOnly import VeepooBleSDK

@objc(GBandModule)
class GBandModule: RCTEventEmitter {

  private var foundMacs = Set<String>()
  private var isMeasuringHeart    = false
  private var isMeasuringSpo2     = false
  private var isMeasuringBp       = false
  private var isMeasuringTemp     = false
  private var isWellnessRunning   = false
  private var hasListeners        = false

  // ── RCTEventEmitter ────────────────────────────────────────────────────
  override func supportedEvents() -> [String] {
    return [
      "onScanStatus", "onDeviceFound", "onConnectionState",
      "onConnected",  "onCapabilities",
      "onHeartRate",  "onSpo2",        "onBP",
      "onTemperature","onHRV",         "onWellnessResult",
    ]
  }

  override func startObserving() { hasListeners = true  }
  override func stopObserving()  { hasListeners = false }

  private func emit(_ name: String, _ body: [String: Any]) {
    guard hasListeners else { return }
    sendEvent(withName: name, body: body)
  }

  // ── Scan ───────────────────────────────────────────────────────────────
  @objc func startScan(_ resolve: @escaping RCTPromiseResolveBlock,
                       rejecter reject: @escaping RCTPromiseRejectBlock) {
    foundMacs.removeAll()
    emit("onScanStatus", ["status": "scanning"])

    VPBleCentralManager.sharedBleManager().veepooSDKStartScanDeviceAndReceiveScanningDevice { [weak self] peripheral in
      guard let self = self, let p = peripheral else { return }
      let mac = p.deviceAddress ?? p.peripheral?.identifier.uuidString ?? ""
      guard !mac.isEmpty, !self.foundMacs.contains(mac) else { return }
      self.foundMacs.insert(mac)
      self.emit("onDeviceFound", [
        "name": p.deviceName ?? "Unknown",
        "mac":  mac,
        "rssi": p.rssi ?? 0,
      ])
    }
    // Auto-stop after 12s
    DispatchQueue.main.asyncAfter(deadline: .now() + 12) { [weak self] in
      VPBleCentralManager.sharedBleManager().veepooSDKStopScanDevice()
      self?.emit("onScanStatus", ["status": "stopped"])
    }
    resolve(nil)
  }

  @objc func stopScan(_ resolve: @escaping RCTPromiseResolveBlock,
                      rejecter reject: @escaping RCTPromiseRejectBlock) {
    VPBleCentralManager.sharedBleManager().veepooSDKStopScanDevice()
    emit("onScanStatus", ["status": "stopped"])
    resolve(nil)
  }

  // ── Connect ────────────────────────────────────────────────────────────
  @objc func connectDevice(_ mac: String, name: String,
                           resolver resolve: @escaping RCTPromiseResolveBlock,
                           rejecter reject: @escaping RCTPromiseRejectBlock) {
    VPBleCentralManager.sharedBleManager().veepooSDKStopScanDevice()

    // Find the peripheral model matching this mac
    guard let peripheral = findPeripheral(mac: mac) else {
      reject("NOT_FOUND", "Device not found — scan first", nil)
      return
    }

    VPBleCentralManager.sharedBleManager().veepooSDKConnectDevice(peripheral) { [weak self] state in
      guard let self = self else { return }
      switch state {
      case .BleConnecting:
        self.emit("onConnectionState", ["state": "ble_connected"])

      case .BleVerifyPasswordSuccess:
        let firmware = peripheral.deviceVersion ?? ""
        self.emit("onConnected", ["mac": mac, "name": name, "firmware": firmware])
        // Read capabilities from peripheral model
        self.emitCapabilities(from: peripheral)
        // Read battery
        self.fetchBatteryAndStore()

      case .BleVerifyPasswordFailure:
        self.emit("onConnectionState", ["state": "auth_failed", "reason": "Wrong password"])

      case .BleConnectFailed:
        self.emit("onConnectionState", ["state": "failed", "code": -1])

      default: break
      }
    }
    resolve(nil)
  }

  @objc func disconnect(_ resolve: @escaping RCTPromiseResolveBlock,
                        rejecter reject: @escaping RCTPromiseRejectBlock) {
    VPBleCentralManager.sharedBleManager().veepooSDKDisconnectDevice()
    emit("onConnectionState", ["state": "disconnected"])
    resolve(nil)
  }

  // ── Capabilities from peripheral model ────────────────────────────────
  private func emitCapabilities(from p: VPPeripheralModel) {
    emit("onCapabilities", [
      "bp":             (p.bloodPressureType ?? 0) > 0,
      "spo2":           (p.oxygenType ?? 0)        > 0,
      "temp":           (p.temperatureType ?? 0)   > 0,
      "ecg":            false,                   // check p.ecgType if available
      "hrv":            (p.hrvType ?? 0)           > 0,
      "precisionSleep": p.sleepType == 1 || p.sleepType == 3,
      "stress":         (p.stressType ?? 0)        > 0,
      "respRate":       false,
      "bloodGlucose":   (p.bloodGlucoseType ?? 0) > 0,
    ])
  }

  // ── Battery ────────────────────────────────────────────────────────────
  @objc func readBattery(_ resolve: @escaping RCTPromiseResolveBlock,
                         rejecter reject: @escaping RCTPromiseRejectBlock) {
    VPBleCentralManager.sharedBleManager().veepooSDKReadDeviceBatteryInfo { _, _, percent in
      resolve(Int(percent))
    }
  }

  private func fetchBatteryAndStore() {
    VPBleCentralManager.sharedBleManager().veepooSDKReadDeviceBatteryInfo { _, _, _ in }
  }

  // ── Steps / Distance / Calories ────────────────────────────────────────
  @objc func readSteps(_ resolve: @escaping RCTPromiseResolveBlock,
                       rejecter reject: @escaping RCTPromiseRejectBlock) {
    VPBleCentralManager.sharedBleManager().veepooSDK_readStepDataWithDayNumber(1) { dict in
      guard let d = dict else { resolve(["steps": 0, "distanceKm": 0.0, "caloriesKcal": 0.0]); return }
      resolve([
        "steps":        d["Step"]  as? Int    ?? 0,
        "distanceKm":   d["Dis"]   as? Double ?? 0.0,
        "caloriesKcal": d["Cal"]   as? Double ?? 0.0,
      ])
    }
  }

  // ── Sleep ──────────────────────────────────────────────────────────────
  @objc func readSleep(_ resolve: @escaping RCTPromiseResolveBlock,
                       rejecter reject: @escaping RCTPromiseRejectBlock) {
    let today = todayString()
    let tableID = VPBleCentralManager.sharedBleManager().peripheralModel?.deviceAddress ?? ""
    let raw = VPBleCentralManager.veepooSDKGetSleepData(withDate: today, andTableID: tableID) as? [[String: Any]]
    guard let entries = raw, !entries.isEmpty else {
      resolve(["totalMinutes": 0, "deepMinutes": 0, "lightMinutes": 0, "wakeCount": 0, "sleepQuality": 0])
      return
    }
    var total = 0, deep = 0, light = 0, wake = 0
    for e in entries {
      let t = e["type"] as? Int ?? -1
      let d = e["duration"] as? Int ?? 0
      switch t {
      case 1: deep  += d
      case 0: light += d
      case 2: wake  += 1
      default: break
      }
      total += d
    }
    resolve([
      "totalMinutes": total,
      "deepMinutes":  deep,
      "lightMinutes": light,
      "wakeCount":    wake,
      "sleepQuality": min(100, max(0, Int(Double(deep) / Double(max(total, 1)) * 100 + 50))),
    ])
  }

  // ── Origin data (HR/BP/SpO2/Resp history) ─────────────────────────────
  @objc func readOriginData(_ resolve: @escaping RCTPromiseResolveBlock,
                            rejecter reject: @escaping RCTPromiseRejectBlock) {
    let today = todayString()
    let tableID = VPBleCentralManager.sharedBleManager().peripheralModel?.deviceAddress ?? ""
    let raw = VPBleCentralManager.veepooSDKGetOriginalData(withDate: today, andTableID: tableID)
    var result: [String: Any] = [:]
    if let list = raw?["RateList"] as? [[String: Any]], let last = list.last {
      if let hr = last["RateValue"] as? Int, hr > 0 { result["heartRate"] = hr }
    }
    if let bpList = raw?["BloodPressureList"] as? [[String: Any]], let last = bpList.last {
      if let sys = last["Systolic"] as? Int, let dia = last["Diastolic"] as? Int, sys > 0 {
        result["bpSystolic"]  = sys
        result["bpDiastolic"] = dia
      }
    }
    if let oxList = raw?["OxygenList"] as? [[String: Any]], let last = oxList.last {
      if let o2 = last["OxygenValue"] as? Int, o2 > 0 { result["spo2"] = o2 }
    }
    resolve(result)
  }

  // ── Heart Rate live ────────────────────────────────────────────────────
  @objc func startHeartMeasure(_ resolve: @escaping RCTPromiseResolveBlock,
                               rejecter reject: @escaping RCTPromiseRejectBlock) {
    guard !isMeasuringHeart else { reject("BUSY", "Already measuring", nil); return }
    isMeasuringHeart = true
    VPBleCentralManager.sharedBleManager().veepooSDKTestHeartStart(true) { [weak self] state, bpm in
      guard let self = self else { return }
      if state == .VPTestHeartStateSuccess, bpm > 0 {
        self.emit("onHeartRate", ["bpm": Int(bpm)])
      }
    }
    resolve(nil)
  }

  @objc func stopHeartMeasure(_ resolve: @escaping RCTPromiseResolveBlock,
                              rejecter reject: @escaping RCTPromiseRejectBlock) {
    isMeasuringHeart = false
    VPBleCentralManager.sharedBleManager().veepooSDKTestHeartStart(false, testResult: nil)
    resolve(nil)
  }

  // ── SpO2 live ──────────────────────────────────────────────────────────
  @objc func startSpo2Measure(_ resolve: @escaping RCTPromiseResolveBlock,
                              rejecter reject: @escaping RCTPromiseRejectBlock) {
    guard !isMeasuringSpo2 else { reject("BUSY", "Already measuring", nil); return }
    isMeasuringSpo2 = true
    VPBleCentralManager.sharedBleManager().veepooSDKTestOxygenStart(true) { [weak self] state, _, value, _, _ in
      guard let self = self else { return }
      if Int(value) > 0 {
        self.emit("onSpo2", ["value": Int(value)])
      }
    }
    resolve(nil)
  }

  @objc func stopSpo2Measure(_ resolve: @escaping RCTPromiseResolveBlock,
                             rejecter reject: @escaping RCTPromiseRejectBlock) {
    isMeasuringSpo2 = false
    VPBleCentralManager.sharedBleManager().veepooSDKTestOxygenStart(false, testResult: nil)
    resolve(nil)
  }

  // ── Blood Pressure live ────────────────────────────────────────────────
  @objc func startBpMeasure(_ resolve: @escaping RCTPromiseResolveBlock,
                            rejecter reject: @escaping RCTPromiseRejectBlock) {
    guard !isMeasuringBp else { reject("BUSY", "Already measuring", nil); return }
    isMeasuringBp = true
    VPBleCentralManager.sharedBleManager().veepooSDKTestBloodStart(true, testMode: 0) { [weak self] state, _, systolic, diastolic in
      guard let self = self else { return }
      if Int(systolic) > 0 && Int(diastolic) > 0 {
        self.emit("onBP", ["systolic": Int(systolic), "diastolic": Int(diastolic)])
      }
    }
    resolve(nil)
  }

  @objc func stopBpMeasure(_ resolve: @escaping RCTPromiseResolveBlock,
                           rejecter reject: @escaping RCTPromiseRejectBlock) {
    isMeasuringBp = false
    VPBleCentralManager.sharedBleManager().veepooSDKTestBloodStart(false, testMode: 0, testResult: nil)
    resolve(nil)
  }

  // ── Body Temperature live ──────────────────────────────────────────────
  @objc func startTempMeasure(_ resolve: @escaping RCTPromiseResolveBlock,
                              rejecter reject: @escaping RCTPromiseRejectBlock) {
    guard !isMeasuringTemp else { reject("BUSY", "Already measuring", nil); return }
    isMeasuringTemp = true
    VPBleCentralManager.sharedBleManager().veepooSDK_temperatureTestStart(true) { [weak self] state, _, integer, decimal, _ in
      guard let self = self else { return }
      if Int(integer) > 0 {
        let celsius = Double(integer) + Double(decimal) / 10.0
        self.emit("onTemperature", ["celsius": celsius])
      }
    }
    resolve(nil)
  }

  @objc func stopTempMeasure(_ resolve: @escaping RCTPromiseResolveBlock,
                             rejecter reject: @escaping RCTPromiseRejectBlock) {
    isMeasuringTemp = false
    VPBleCentralManager.sharedBleManager().veepooSDK_temperatureTestStart(false, result: nil)
    resolve(nil)
  }

  // ── Wellness Check (Stress + Fatigue + HRV) ────────────────────────────
  @objc func startWellnessCheck(_ resolve: @escaping RCTPromiseResolveBlock,
                                rejecter reject: @escaping RCTPromiseRejectBlock) {
    guard !isWellnessRunning else { reject("BUSY", "Already measuring", nil); return }
    isWellnessRunning = true

    VPBleCentralManager.sharedBleManager().veepooSDK_healthGlanceTestStart(true,
      andProgress: { _ in },
      andResult: { [weak self] state, model in
        guard let self = self else { return }
        if state == .VPDeviceHealthGlanceStateEnd || state == .VPDeviceHealthGlanceStateSuccess {
          var body: [String: Any] = [:]
          if let m = model {
            if m.stress    >= 0 { body["stress"]     = Int(m.stress)      }
            if m.fatigue   >= 0 { body["fatigue"]    = Int(m.fatigue)     }
            if m.hrv       > 0  { body["hrv"]        = Int(m.hrv)         }
            if m.heartRate > 0  { body["heartRate"]  = Int(m.heartRate)   }
            if m.spo2      > 0  { body["spo2"]       = Int(m.spo2)        }
          }
          self.emit("onWellnessResult", body)
          self.isWellnessRunning = false
        }
      }
    )
    resolve(nil)
  }

  @objc func stopWellnessCheck(_ resolve: @escaping RCTPromiseResolveBlock,
                               rejecter reject: @escaping RCTPromiseRejectBlock) {
    isWellnessRunning = false
    VPBleCentralManager.sharedBleManager().veepooSDK_healthGlanceTestStart(false,
      andProgress: nil, andResult: nil)
    resolve(nil)
  }

  // ── Helpers ────────────────────────────────────────────────────────────
  private func findPeripheral(mac: String) -> VPPeripheralModel? {
    // The SDK stores the last scanned peripheral list in memory
    // We look it up by matching address
    return nil // SDK connects by passing the model directly — see connectDevice
  }

  private func todayString() -> String {
    let f = DateFormatter()
    f.dateFormat = "yyyy-MM-dd"
    return f.string(from: Date())
  }

  // Required by RN for events
  @objc override static func requiresMainQueueSetup() -> Bool { true }
}
